File: hw4/problem_5.py
import numpy as np
import csv
import pandas as pd
from collections import defaultdict
import igraph
from pprint import pprint
from igraph import *


def find_euler_tour(graph):
    tour = []
    E = graph

    numEdges = defaultdict(int)

    def find_tour(u):
        for e in E:
            if u == e[0]:
                u,v = e
                E.remove(e)
                find_tour(v)
            elif u == e[1]:
                v,u = e
                E.remove(e)
                find_tour(v)
        tour.insert(0,u)

    for i,j in graph:
        numEdges[i] += 1
        numEdges[j] += 1

    start = graph[0][0]
    for i,j in numEdges.iteritems():
        if j % 2 > 0:
            start = i
            break

    current = start
    find_tour(current)

    if tour[0] != tour[-1]:
        return None
    return tour


g = Graph.Read_Ncol('./finance_data/dup_mst.txt', directed=False, weights = True)
el = g.get_edgelist()

tour = find_euler_tour(el)

tsp = []
tsp_name = []

for i in tour:
    if i not in tsp:
        tsp.append(i)
        tsp_name.append(g.vs[i]['name'])
tsp_name.append(g.vs[0]['name'])
#we need to go back to the first node
'''
f = open('./finance_data/tsp_name.txt', 'w')
for item in tsp_name:
    print >> f, item

f.close()
'''
ff = open('./finance_data/tsp_all.txt', 'w')
for i in tour:
    print >> ff, g.vs[i]['name']
ff.close()
    

File: hw4/problem_5.R
library("igraph")
require(graph)
require(eulerian)
library("TSP")

adj_matrix = read.csv("/Users/kay/Documents/course/ee232/hw_4/finance_data/adj_matrix.csv")
adj_matrix$X = NULL
adj_matrix = as.matrix(adj_matrix)
g = graph_from_adjacency_matrix(adj_matrix, mode = "undirected", weighted = TRUE, add.colnames = NULL, add.rownames = NA)
mst_g = mst(g, weights = E(g)$weight)

for(v1 in V(g))
{
  for(v2 in V(g))
  {
    for(v3 in V(g))
    {
      if(v1!=v2 && v1!=v3 && v2!=v3)
      {
        if(D[v1, v2] > D[v1, v3] + D[v2, v3])
        {
          print("the triangle inequality does not hold")
          break
        }
        else
          next
      }
    }
    break
  }
  break
}

mst_directed = as.directed(mst_g, 'mutual')
double_mst = as.undirected(mst_directed, 'each' )

el = get.edgelist(double_mst)
v3 = as.vector(E(double_mst)$weight)
dup = cbind(el, v3)

#write the duplicated mst graph to file
write.table(dup, file = "/Users/kay/Documents/course/ee232/hw_4/finance_data/dup_mst.txt",row.names=FALSE, col.names = FALSE)

#Read eulerian result from python 
tour_path = read.table("/Users/kay/Documents/course/ee232/hw_4/finance_data/tsp_name.txt")
tsp_path_name = as.vector(tour_path[,1])

node_name = V(g)$name
tsp_path_index = numeric()
for(i in 1:length(tsp_path_name)){
  match_index = which(tsp_path_name[i] == node_name)
  tsp_path_index = c(tsp_path_index, match_index)
}

tsp_weight = 0
for(i in 2:length(tsp_path_index)){
  tsp_weight = tsp_weight + adj_matrix[tsp_path_index[i],tsp_path_index[i-1]]
}

sum(E(mst_g)$weight)
#417.819
sum(E(double_mst)$weight)
#835.638
tsp_weight
#477.1329

#use a general tsp package
etsp = ETSP(adj_matrix)
tour = solve_TSP(etsp)
tour
#tour length: 768.3892 
plot(etsp, tour)
